### PR TITLE
joel/cache-key-callback

### PR DESCRIPTION
**Pull request by JoelAlphonso**
This pull request adds a callback function in the container that can be overriden in a project to add extra processing on the middle cache key. This allows to add extra modifiers to it la session variables for example

### CODE REVIEW

File: src/Charcoal/Cache/Middleware/CacheMiddleware.php [hunk 1]
```diff
@@ -131,7 +138,7 @@ public function __construct(array $data)
         $this->excludedQuery = $data['excluded_query'];
         $this->ignoredQuery  = $data['ignored_query'];
 
-        $this->skipCache = (array)$data['skip_cache'];
```
Comment (mcaskill) [inline src/Charcoal/Cache/Middleware/CacheMiddleware.php:134]: I'm unfamiliar with this property. Why is it removed but the default value in `defaults()` preserved?

File: src/Charcoal/Cache/Middleware/CacheMiddleware.php [hunk 2]
```diff
@@ -265,7 +274,10 @@ private function cacheKeyFromRequest(RequestInterface $request)
         }
 
         $cacheKey = 'request/' . $request->getMethod() . '/' . md5((string)$uri);
-        return $cacheKey;
+
+        $callback = $this->processCacheKeyCallback;
+
+        return $callback($cacheKey);
```
Comment (mcaskill) [inline src/Charcoal/Cache/Middleware/CacheMiddleware.php]:
```suggestion
        $callback = $this->processCacheKeyCallback;
        if (is_callable($callback)) {
            return $callback($cacheKey);
        }

        return $cacheKey;
```

File: src/Charcoal/Cache/ServiceProvider/CacheServiceProvider.php [hunk 3]
```diff
@@ -267,6 +267,17 @@ public function registerService(Container $container)
...
+        /**
+         * Provides a way to add custom processing over the cache key for the cache middleware.
+         *
+         * @return Closure
+         */
+        $container['cache/process-cache-key-callback'] = function () {
+            return function ($key) {
+                return $key;
+            };
+        };
```
Comment (mcaskill) [inline src/Charcoal/Cache/ServiceProvider/CacheServiceProvider.php]: Rename service parameter to `cache/middleware/process-cache-key-callback`.
Maybe remove this and implement the suggestion in `CacheMiddleware.php` where if the callback is missing, return the key as is.

**Review by mcaskill — CHANGES_REQUESTED** [review]
What if:

```php
        /**
         * The cache middleware configset.
         *
         * @param  Container $container The service container.
         * @return array
         */
        $container['cache/middleware/config'] = function (Container $container) {
            $appConfig = isset($container['config']) ? $container['config'] : [];

            if (isset($appConfig['middlewares']['charcoal/cache/middleware/cache'])) {
                $wareConfig = $appConfig['middlewares']['charcoal/cache/middleware/cache'];
            } else {
                $wareConfig = [];
            }

            return $wareConfig;
        };

        /**
         * The middleware for caching HTTP responses.
         *
         * @param  Container $container A container instance.
         * @return CacheMiddleware
         */
        $container['middlewares/charcoal/cache/middleware/cache'] = function (Container $container) {
            return new CacheMiddleware($wareConfig + $container['cache/middleware/config']);
        };
```

And then in projects:

```php
$container->extend('cache/middleware/config', function (array $config) {
    $config['processCacheKeyCallback'] = function ($key) {
        return strtoupper($key);
    };
    return $config;
});
```

Or add a method `CacheMiddleware::setProcessCacheKeyCallback()`:

```php
$container->extend('middlewares/charcoal/cache/middleware/cache', function (CacheMiddleware $middleware) {
    $middleware->setProcessCacheKeyCallback(function ($key) {
        return strtoupper($key);
    });
    return $middleware;
});
```

File: src/Charcoal/Cache/Middleware/CacheMiddleware.php (hunk 1 above)
Comment (JoelAlphonso) [reply src/Charcoal/Cache/Middleware/CacheMiddleware.php:134]: oups

File: src/Charcoal/Cache/ServiceProvider/CacheServiceProvider.php (hunk 3 above)
Comment (mcaskill) [inline src/Charcoal/Cache/ServiceProvider/CacheServiceProvider.php]: Appears to be obsolete.

**Review by mcaskill — CHANGES_REQUESTED** [review]

File: src/Charcoal/Cache/ServiceProvider/CacheServiceProvider.php (hunk 3 above)
Comment (JoelAlphonso) [reply src/Charcoal/Cache/ServiceProvider/CacheServiceProvider.php]: Indeed, je croyais l'avoir enlevé thanks

**Pull request merged**